www/blog/2022-11-23-ts-perf.md

```diff
@@ -0,0 +1,192 @@
+---
+slug: ts-perf
+title: TypeScript performance lessons while refactoring for v10
+author: Sachin Raja
+author_title: tRPC Core Team Member
+author_url: https://twitter.com/s4chinraja
+author_image_url: https://avatars1.githubusercontent.com/u/58836760?s=460&v=4
+---
+
+As library authors, our goal is to provide the best possible developer experience (DX) for our peers. Reducing time-to-error and providing intuitive APIs removes mental overhead from the minds of developers so that they can focus on what's most important: great end user experience.
+
+---
+
+It's no secret that TypeScript is the driving force behind how tRPC ships DX. TypeScript adoption is the modern standard in delivering great JavaScript-based experiences today - but this improved certainty around types does have some tradeoffs.
+
+Today, the TypeScript type checker is prone to becoming slow (although releases like [TS 4.9](https://devblogs.microsoft.com/typescript/announcing-typescript-4-9/#performance-improvements) are promising!). Libraries almost always contain the fanciest TypeScript incantations in your codebase, pressing your TS compiler to its limits. For this reason, library authors like us must be mindful of our contributions to that burden and do our best to keep your IDE working as fast as possible.
+
+## Automating library performance
+
+While tRPC was in `v9`, we began seeing reports from developers that their large tRPC routers were starting to have detrimental effects on their type checker. This was a new experience for tRPC as we saw tremendous adoption during the `v9` phase of tRPC's development. With more developers creating larger and larger products with tRPC, some cracks began to show.
+
+Your library may not be slow _now_, but it's important to keep an eye on performance as your library grows and changes. Automated testing can remove an immense burden from your library authoring (and application building!) by _programatically_ testing your library code on each commit.
+
+For tRPC, we ensure this by [generating](https://github.com/trpc/trpc/blob/9fc2d06a8924da73e10b9d4497f3a1f53de706ed/scripts/generateBigBoi.ts) and [testing](https://github.com/trpc/trpc/blob/9fc2d06a8924da73e10b9d4497f3a1f53de706ed/packages/tests/server/react/bigBoi.test.tsx) a router with 3,500 procedures and 1,000 routers. We test all three pieces of the library (server, vanilla client, and the React client) because they all have different code paths. In the past, we have seen regressions that are isolated to one section of the library and rely on our tests to show us when those unexpected behaviors occur.
```

Comment:
Maybe add some more:

```suggestion
For tRPC, we ensure this by [generating](https://github.com/trpc/trpc/blob/9fc2d06a8924da73e10b9d4497f3a1f53de706ed/scripts/generateBigBoi.ts) and [testing](https://github.com/trpc/trpc/blob/9fc2d06a8924da73e10b9d4497f3a1f53de706ed/packages/tests/server/react/bigBoi.test.tsx) a router with 3,500 procedures and 1,000 routers. We test all three pieces of the library (server, vanilla client, and the React client) because they all have different code paths. In the past, we have seen regressions that are isolated to one section of the library and rely on our tests to show us when those unexpected behaviors occur. (We still do [want to do more](https://github.com/trpc/trpc/issues/2892) to measure compilation times)
```